control-plane/gateways/builder.go

```diff
@@ -5,21 +5,31 @@ package gateways
 
 import (
 	meshv2beta1 "github.com/hashicorp/consul-k8s/control-plane/api/mesh/v2beta1"
+	corev1 "k8s.io/api/core/v1"
 )
 
-// meshGatewayBuilder is a helper struct for building the Kubernetes resources for a mesh gateway.
+type Gateway interface {
+	*meshv2beta1.MeshGateway | *meshv2beta1.APIGateway
+	GetName() string
+	GetNamespace() string
+	ListenersToPorts(int32) []corev1.ServicePort
+	GetAnnotations() map[string]string
+	GetLabels() map[string]string
+}
+
+// gatewayBuilder is a helper struct for building the Kubernetes resources for a mesh gateway.
 // This includes Deployment, Role, Service, and ServiceAccount resources.
 // Configuration is combined from the MeshGateway, GatewayConfig, and GatewayClassConfig.
-type meshGatewayBuilder struct {
-	gateway *meshv2beta1.MeshGateway
-	config  GatewayConfig
+type gatewayBuilder[T Gateway] struct {
```

Comment:
the generic here allows us to restrict  this to only gateway types, I wasn't a huge fan of making all these use getters in the code, but the current generics implementation doesn't allow for struct fields. the other option I considered was having the `gatewayBuilder` take the name/namespace/annotations/labels as arguments to the builder rather than passing the entire gateway, but there is the issue of the listeners where the behavior of converting listeners to ports for mesh gateways is slightly different from api gateways.